lighthouse-core/audits/resource-budget.js

```diff
@@ -0,0 +1,182 @@
+/**
+ * @license Copyright 2019 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const Audit = require('./audit.js');
+const ResourceSummary = require('../computed/resource-summary.js');
+const i18n = require('../lib/i18n/i18n.js');
+
+const UIStrings = {
+  /** Imperative title of a Lighthouse audit that tells the user to minimize the size and quantity of resources used to load the page. */
+  title: 'Keep request counts low and file sizes small',
+  /** Description of a Lighthouse audit that tells the user that they can setup a budgets for the quantity and size of page resources. No character length limits. */
+  description: 'To set budgets for the quantity and size of page resources,' +
+    ' add a budget.json file.',
+  /** [ICU Syntax] Label identifying the number of requests*/
+  requestCount: `{count, plural,
+    =1 {1 request}
+    other {# requests}
+   }`,
+};
+
+const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+
+/** @typedef {{count: number, size: number}} ResourceEntry */
+
+class ResourceBudget extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'resource-budget',
```

Comment:
this name indicates that 'metric-budget' may be a different audit.
i'm thinking about users who are consuming the JSON and want to fail a build based on if the perf budget was violated. if 'metric-budget' was separate, they'd have to also check that second audit and its score.

the alternative I see is that we have a single 'performance-budget' audit that will have all results coming from the budgets.
however, i dont exactly know the shape of `details` that we'd have if both resource + metric budget results are in there, and yet we expect to render them quite differently. :/ @brendankenny 

thoughts y'all?

Reply:
> thoughts y'all?

yeah, I think calling it `performance-budget` and having it include resources and metrics makes total sense. I'd want a single spot in the json to check my budgets as well.

The columns might get excessive in the table, but we can figure that out :) e.g. put more smarts in `performance-category-renderer.js` and have it render two tables from one set of audit-details